### DISCUSSION TITLE
Real time performace impacted by ethernet traffic

### DISCUSSION THREAD

**Discussion by aneejosep** [zephyrproject-rtos/zephyr#75424] [General] [answered]
I am evaluating Zephyr for one of the upcoming product development and found that the real time performance is greatly affected when there is traffic in the network interface. 

The way I do my test is as follows.

- A timer to signal a semaphore every 1ms.
- A high priority task to receive this signal and add trace point.
- Adding trace point is nothing but to get the current time and compute the difference from the previous point (ideally it should be 1ms) and store it in an array.
- Record 100K points and after tests are done, print the values to console.
- Use these values to plot a histogram.

I see that when no network traffic the results are ok. But when there is network traffic, the values even shift more than 500μs on an STM32 board. On an NXP board it is even worse like 3+ ms jitter.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/10137311/cb0f31ac-5317-4a74-9114-f87478d277f6)

I tried to search for configuring the network priority, but I could not find any useful information.
Am I missing something here?

**Comment by jukkar** [comment]
If co-operative threads are enabled, then by default network threads are run in co-operative mode.
You can tweak this by setting `CONFIG_NET_TC_THREAD_PREEMPTIVE=y` which allow system to preempt network handler threads. This could help your perf numbers.
In both cases the network thread priorities can be tweaked too if needed.

**Comment by aneejosep** [comment]
Hello @jukkar thanks for the response.
I did the setting CONFIG_NET_TC_THREAD_PREEMPTIVE=y as you said. It improved a little bit but still the jitter is around 200μs for 1ms task when there is Ethernet traffic.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/10137311/b23ec6ef-4033-44b6-be8d-6ea47eca6882)

In our application, if the jitter of the real time task is above 100μs, it is not acceptable. Why ethernet is causing a lot of jitter in the realtime task? is it because of the Ethernet generating a lot of interrupts?

**Reply by jukkar** [reply]
Difficult to say, I am not very familiar with what the vendor Ethernet driver is actually doing.

You can also experiment with the TX/RX thread priority by setting
`CONFIG_NET_TC_THREAD_PRIO_CUSTOM=y`
`CONFIG_NET_TC_TX_THREAD_BASE_PRIO=<new-priority>`
`CONFIG_NET_TC_RX_THREAD_BASE_PRIO=<new-priority>`
The default value for the base priority is 0, so you can experiment with higher value (>0).

**Reply by yashi** [reply, marked as answer]
I'm not sure this is the one but eth_stm32_hal.c has the following block in `rx_thread()`

```c
			while ((pkt = eth_rx(dev)) != NULL) {
				iface = net_pkt_iface(pkt);
#if defined(CONFIG_NET_DSA)
				iface = dsa_net_recv(iface, &pkt);
#endif
				res = net_recv_data(iface, pkt);
				if (res < 0) {
					eth_stats_update_errors_rx(
							net_pkt_iface(pkt));
					LOG_ERR("Failed to enqueue frame "
						"into RX queue: %d", res);
					net_pkt_unref(pkt);
				}
			}
```

This means that when the device is kept receiving packetes, it won't leave the function. The thread which calls this function is created as:

```c
		/* Start interruption-poll thread */
		k_thread_create(&dev_data->rx_thread, dev_data->rx_thread_stack,
				K_KERNEL_STACK_SIZEOF(dev_data->rx_thread_stack),
				rx_thread, (void *) dev, NULL, NULL,
				K_PRIO_COOP(CONFIG_ETH_STM32_HAL_RX_THREAD_PRIO),
				0, K_NO_WAIT);
```

`CONFIG_ETH_STM32_HAL_RX_THREAD_PRIO` is default to `2`.

**Reply by aneejosep** [reply]
@yashi thanks for the answer.
I modified the thread creation to use priority +2(preemptive) and the issue is gone.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/10137311/2a59343c-0c40-47bf-9f91-c9dc99d27495)

Now I need to find whats wrong with nxp driver.

**Reply by aneejosep** [reply]
NXP has fixed the issue with the [pull request](https://github.com/zephyrproject-rtos/zephyr/pull/75625 ) 

The issue in STM32 HAL also should be raised as a bug? i think it is not the right way to create a cooperative thread directly from HAL like that. At least it should be configurable by the user.

**Reply by jukkar** [reply]
> The issue in STM32 HAL also should be raised as a bug? i think it is not the right way to create a cooperative thread directly from HAL like that. At least it should be configurable by the user.

Sure, please create an issue for that.

**Comment by DerekSnell** [comment]
Thank you, @aneejosep 
To help others optimize their application, we wanted to share additional details that helped improve the performance here on the NXP MIMXRT1170-EVK board in addition to fixing a bug in the Ethernet driver https://github.com/zephyrproject-rtos/zephyr/pull/75625.

Changing thread priorities can help.  These changes set the networking and system work queue threads to be preemptive, and allow an application thread to preempt them.
```
CONFIG_NET_TC_THREAD_PREEMPTIVE=y
CONFIG_SYSTEM_WORKQUEUE_PRIORITY=3
CONFIG_NET_TCP_WORKER_PRIO=0
```

Executing from internal ITCM SRAM also provides a major performance boost over external QSPI flash.  The entire application can be moved to ITCM in devicetree by setting the chosen node `zephyr,flash = &itcm`.  This option requires loading the code into ITCM before the app executes.  That can be done with a debugger, or by configuring the bootloader in ROM.  Another option is to relocate specific files, libraries, or functions to ITCM, and leave the rest to execute from flash.  One option for this is using Zephyr's [Code Relocation](https://docs.zephyrproject.org/latest/kernel/code-relocation.html) feature and selecting `CONFIG_CODE_DATA_RELOCATION=y`.  Then the application `CMakeLists.txt` file can specify specific files and libraries to relocate.  Listed below are some files relocated to improve performance.  Likely, not all of these are needed, but the optimization is application-specific, so these are provided as an example.
```
if (CONFIG_CODE_DATA_RELOCATION)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/sched.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/kheap.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/mem_slab.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/msg_q.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/mutex.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/queue.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/sem.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/system_work_q.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/work.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/thread.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/timeslicing.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/thread_monitor.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/timeout.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/poll.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/usage.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/mempool.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/busy_wait.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/device.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/idle.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/kernel/timer.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/heap.c LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(LIBRARY arch__arm__core LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(LIBRARY arch__arm__core__mpu LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(LIBRARY drivers__timer LOCATION ITCM_TEXT_RODATA NOKEEP)
    zephyr_code_relocate(LIBRARY drivers__timer LOCATION ITCM_TEXT_RODATA NOKEEP)

# Relocate key networking stack components and L2 layer to RAM
  zephyr_code_relocate(LIBRARY subsys__net__ip LOCATION ITCM_TEXT_RODATA NOKEEP)
  zephyr_code_relocate(LIBRARY subsys__net LOCATION ITCM_TEXT_RODATA NOKEEP)
  zephyr_code_relocate(LIBRARY subsys__net__lib__config LOCATION ITCM_TEXT_RODATA NOKEEP)
if (CONFIG_NET_L2_ETHERNET)
  zephyr_code_relocate(LIBRARY drivers__ethernet LOCATION ITCM_TEXT_RODATA NOKEEP)
  zephyr_code_relocate(LIBRARY drivers__mdio LOCATION ITCM_TEXT_RODATA NOKEEP)
  zephyr_code_relocate(LIBRARY subsys__net__l2__ethernet LOCATION ITCM_TEXT_RODATA NOKEEP)
  zephyr_code_relocate(FILES $ENV{ZEPHYR_BASE}/../modules/hal/nxp/mcux/mcux-sdk/drivers/enet/fsl_enet.c
    LOCATION ITCM_TEXT_RODATA NOKEEP)
endif()
endif()
```

**Reply by mikk-leini-krakul** [reply]
Could you also explain how to move ISR vector tables into ITCM (or DTCM, but ISR's into ITCM)?
CONFIG_SRAM_VECTOR_TABLE moves them to SRAM (OCRAM).

**Comment by marwaiehm-st** [comment]
The issue in STM32 HAL is fixed with this [pull request](https://github.com/zephyrproject-rtos/zephyr/pull/77904)

**Comment by maass-hamburg** [comment]
Btw if you would use the generic zero copy dwc_mac driver and that together with https://github.com/zephyrproject-rtos/zephyr/pull/114084 you should get even better performance, as as all the rx is happening in the interrupt and the dma is already directly copying the date in the correct net_buf.
